lib/build.py

```diff
@@ -235,28 +248,29 @@ def __default__(self, tree):
          inst.init(self.inst_prev)
          # The three announce_maybe() calls are clunky but I couldn’t figure
          # out how to avoid the repeats.
-         try:
-            self.miss_ct = inst.prepare(self.miss_ct)
-            inst.announce_maybe()
-         except Instruction_Ignored:
-            inst.announce_maybe()
-            return
-         except ch.Fatal_Error:
-            inst.announce_maybe()
-            raise
-         if (inst.miss):
-            if (self.miss_ct == 1):
-               inst.checkout_for_build()
-            try:
-               inst.execute()
-            except ch.Fatal_Error:
-               inst.rollback()
-               raise
-            if (inst.image_i >= 0):
-               inst.metadata_update()
-            inst.commit()
-         self.inst_prev = inst
-         self.instruction_total_ct += 1
+         if (stop_build is None):
+           try:
+              self.miss_ct = inst.prepare(self.miss_ct)
+              inst.announce_maybe()
+           except Instruction_Ignored:
+              inst.announce_maybe()
+              return
+           except ch.Fatal_Error:
+              inst.announce_maybe()
+              raise
+           if (inst.miss):
+              if (self.miss_ct == 1):
+                 inst.checkout_for_build()
+              try:
+                 inst.execute()
+              except ch.Fatal_Error:
+                 inst.rollback()
+                 raise
+              if (inst.image_i >= 0):
+                 inst.metadata_update()
+              inst.commit()
+           self.inst_prev = inst
+           self.instruction_total_ct += 1
```

Comment:
What if you literally prune the tree to remove everything after the last stage we want? Then none of the special cases would be needed at all.